relay-server/src/actors/project_cache.rs

```diff
@@ -86,16 +86,26 @@ impl ProjectCache {
     fn get_or_create_project(&mut self, project_key: ProjectKey) -> &mut Project {
         metric!(histogram(RelayHistograms::ProjectStateCacheSize) = self.projects.len() as u64);
 
-        let config = self.config.clone();
+        let config = &self.config;
 
         self.projects
             .entry(project_key)
-            .and_modify(|_| {
+            .and_modify(|project| {
+                if let Some(state) = project.state() {
+                    if let Expiry::Expired = state.check_expiry(config) {
+                        // Treat expired states as absent by mimicking the "create" branch here.
+                        metric!(counter(RelayCounters::ProjectCacheMiss) += 1);
+
+                        // NOTE: This drops the old project. We might want to defer dropping it
+                        // to `evict_stale_project_caches` instead (e.g. by swapping it to a garbage list).
+                        *project = Project::new(project_key, Arc::clone(config));
+                    }
```

Comment:
Open question: Cache eviction actually uses a different condition to drop the project. Note the factor `2` and usage of `last_updated` instead of `last_fetch`. Should we use that same condition here?

https://github.com/getsentry/relay/blob/3fa18443e31ef6cfd6d0f47d46a7d32675874120/relay-server/src/actors/project_cache.rs#L78-L81

Reply:
Yes. Ideally, the logic in both places is the same -- we don't want to treat a project as expired in one place, and not do that in another. Further, it'd be great if we also share the same logic to evict a project as you suggest in the comment.

Reply:
Talked to Jan about this: it's OK if the logic for expiry stays separate.

Reply:
Let's put the reasoning for this in a code comment, we'll trip over this in the future again. 

I agree with Iker here, the logic for whether a project is expired or not should be shared, ideally. For eviction, we want to add an _additional_ delay to avoid prematurely deleting a project that is just being fetched. That is, a project may just have become expired, but a background fetch is already going. In that case, we should keep it in the cache. This is why the formula here has the `2 *` in the delta computation.

Reply:
Added documentation to `evict_stale_project_caches`:

https://github.com/getsentry/relay/blob/1d0b98df1f0480b31fe76890b9364b1cdd60d035/relay-server/src/actors/project_cache.rs#L75-L80